src/Orleans/Streams/PubSub/MemoryPubSubStore.cs

```diff
@@ -0,0 +1,258 @@
+            consumers = new ConcurrentDictionary<PubSubSubscriptionState, byte>();
+        }
+
+        public void AddProducer(PubSubPublisherState state)
+        {
+            producers.AddOrUpdate(state, byte.MinValue, (publisherState, b) => b);
+        }
+
+        public void RemoveProducer(PubSubPublisherState state)
+        {
+            byte b;
+            bool re = producers.TryRemove(state, out b);
+            if (!re)
+            {
+                throw new OrleansException($"{this.GetType().Name} : Remove producer failed");
+            }
+        }
+
+        public void AddConsumer(PubSubSubscriptionState state)
+        {
+            consumers.AddOrUpdate(state, byte.MinValue, (subscriptionState, b) => b);
+        }
+
+        public void RemoveConsumer(PubSubSubscriptionState state)
+        {
+            byte b;
+            bool re = consumers.TryRemove(state, out b);
+            if (!re)
+            {
+                throw new OrleansException($"{this.GetType().Name} : Remove consumer failed");
+            }
+        }
+
+        public ICollection<PubSubPublisherState> Producers => producers.Keys;
+        public ICollection<PubSubSubscriptionState> Consumers => consumers.Keys;
+    }
+    
+    internal class MemoryPubSubStore
+    {
+        private readonly ConcurrentDictionary<StreamId, MemoryPubSubStreamSubscriptions> store;
+        private readonly Logger logger;
+        public MemoryPubSubStore()
+        {
+            store = new ConcurrentDictionary<StreamId, MemoryPubSubStreamSubscriptions>();
+            logger = LogManager.GetLogger(this.GetType().Name, LoggerType.Runtime);
+        }
+
+        public ISet<PubSubSubscriptionState> RegisterProducer(StreamId streamId, string streamProvider,
+            IStreamProducerExtension streamProducer)
+        {
+            var publisherState = new PubSubPublisherState(streamId, streamProducer);
+            MemoryPubSubStreamSubscriptions subscriptions = new MemoryPubSubStreamSubscriptions();
+            try
+            {
+                subscriptions.AddProducer(publisherState);
+                subscriptions = store.AddOrUpdate(streamId, subscriptions, (id, streamSubscriptions) =>
+                {
+                    streamSubscriptions.AddProducer(publisherState);
+                    return streamSubscriptions;
+                });
+            }
+            catch (Exception exc)
+            {
+                logger.Error(ErrorCode.Stream_RegisterProducerFailed, $"Failed to register a stream producer.  Stream: {streamId}, Producer: {streamProducer}", exc);
+                throw;
+            }
+            return subscriptions.Consumers.Where(c => !c.IsFaulted).ToSet();
+        }
+
+        public void UnregisterProducer(StreamId streamId, string streamProvider, IStreamProducerExtension streamProducer)
+        {
+            var publisherState = new PubSubPublisherState(streamId, streamProducer);
```

Comment:
in general we want to delay allocation until needed.  This could be deferred until the store.TryGetValue check, correct?